src/NetTopologySuite/Geometries/CoordinateSequences.cs

```diff
@@ -54,10 +56,179 @@ public static void Swap(CoordinateSequence seq, int i, int j)
         /// <param name="length">The number of coordinates to copy</param>
         public static void Copy(CoordinateSequence src, int srcPos, CoordinateSequence dest, int destPos, int length)
         {
+            // Attempt shortcuts
+            if (src is PackedDoubleCoordinateSequence srcPD && dest is PackedDoubleCoordinateSequence destPD)
+            {
+                if (TryRawCopy(srcPD, srcPos, destPD, destPos, length))
+                    return;
+            }
+            else if (src is PackedFloatCoordinateSequence srcPF && dest is PackedFloatCoordinateSequence destPF)
+            {
+                if (TryRawCopy(srcPF, srcPos, destPF, destPos, length))
+                    return;
+            }
+            else if (src is DotSpatialAffineCoordinateSequence srcDS && dest is DotSpatialAffineCoordinateSequence destDS)
+            {
+                if (TryRawCopy(srcDS, srcPos, destDS, destPos, length))
+                    return;
+            }
+
+            GetCommonOrdinateIndices(src, dest, out int[] srcIndices, out int[] destIndices);
+
+            // Copy one by one
             for (int i = 0; i < length; i++)
-                CopyCoord(src, srcPos + i, dest, destPos + i);
+                CopyCoord(src, srcPos + i, srcIndices, dest, destPos + i, destIndices);
         }
 
+        /// <summary>
+        /// Get the common ordinate indices of two <c>CoordinateSequence</c>s.
+        /// </summary>
+        /// <param name="seq0">A <c>CoordinateSequence</c></param>
+        /// <param name="seq1">A <c>CoordinateSequence</c></param>
+        /// <param name="seq0Indices">The array of common ordinate indices as in <paramref name="seq0"/></param>
+        /// <param name="seq1Indices">The array of common ordinate indices as in <paramref name="seq1"/></param>
+        private static void GetCommonOrdinateIndices(CoordinateSequence seq0, CoordinateSequence seq1,
+            out int[] seq0Indices, out int[] seq1Indices)
```

Comment:
I think we shouldn't need this method.  The revamped `CoordinateSequence` class is designed so that you can scan very efficiently through its coordinates and dimensions, in order, with simple `for` loops like the ones in my bigger comment in this same method.  No need to try to extract separate lists of things that the two have in common.

In other words, every pair of sequences will have the first 2 spatial dimensions in common, and then the next 0..n spatial dimensions in common until one of them runs out of spatial dimensions, and then 0..n measures in common until one of them runs out of measures (the "measured" dimension might not have the same index on both sides).